drivers/usb/usbehci_new/roothub.c

```diff
@@ -0,0 +1,710 @@
+
+    *PortStatus = status;
+
+    if (status.PortStatus.Usb20PortStatus.CurrentConnectStatus)
+    {
+        DPRINT_RH("EHCI_RH_GetPortStatus: Port - %x, status.AsULONG - %X\n",
+                  Port,
+                  status.AsUlong32);
+    }
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+EHCI_RH_GetHubStatus(IN PVOID ehciExtension,
+                     IN PUSB_HUB_STATUS_AND_CHANGE HubStatus)
+{
+    DPRINT_RH("EHCI_RH_GetHubStatus: ... \n");
+    HubStatus->AsUlong32 = 0;
+    return MP_STATUS_SUCCESS;
+}
+
+VOID
+NTAPI
+EHCI_RH_FinishReset(IN PVOID ehciExtension,
+                    IN PVOID Context)
+{
+    PEHCI_EXTENSION EhciExtension = ehciExtension;
+    PULONG PortStatusReg;
+    EHCI_PORT_STATUS_CONTROL PortSC;
+    PUSHORT Port = Context;
+
+    DPRINT("EHCI_RH_FinishReset: *Port - %x\n", *Port);
+
+    PortStatusReg = &EhciExtension->OperationalRegs->PortControl[*Port - 1].AsULONG;
+    PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+
+    if (PortSC.AsULONG != -1)
+    {
+        if (!PortSC.CurrentConnectStatus)
+            DPRINT("EHCI_RH_FinishReset: PortSC.AsULONG - %X\n", PortSC.AsULONG);
+
+        if (PortSC.PortEnabledDisabled ||
+            !PortSC.CurrentConnectStatus ||
+            PortSC.ConnectStatusChange)
+        {
+            EhciExtension->FinishResetPortBits |= (1 << (*Port - 1));
+            RegPacket.UsbPortInvalidateRootHub(EhciExtension);
+        }
+        else
+        {
+            PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+            PortSC.PortOwner = EHCI_PORT_OWNER_COMPANION_CONTROLLER;
+            WRITE_REGISTER_ULONG(PortStatusReg, PortSC.AsULONG);
+            EhciExtension->FinishResetPortBits |= (1 << (*Port - 1));
+        }
+
+        EhciExtension->ResetPortBits &= ~(1 << (*Port - 1));
+    }
+}
+
+VOID
+NTAPI
+EHCI_RH_PortResetComplete(IN PVOID ehciExtension,
+                          IN PVOID Context)
+{
+    PEHCI_EXTENSION EhciExtension = ehciExtension;
+    PULONG PortStatusReg;
+    EHCI_PORT_STATUS_CONTROL PortSC;
+    ULONG ix;
+    PUSHORT Port = Context;
+
+    DPRINT("EHCI_RH_PortResetComplete: *Port - %x\n", *Port);
+
+    PortStatusReg = &EhciExtension->OperationalRegs->PortControl[*Port - 1].AsULONG;
+
+    do
+    {
+        PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+
+        PortSC.ConnectStatusChange = 0;
+        PortSC.PortEnableDisableChange = 0;
+        PortSC.OverCurrentChange = 0;
+        PortSC.PortReset = 0;
+
+        WRITE_REGISTER_ULONG(PortStatusReg, PortSC.AsULONG);
+
+        for (ix = 0; ix <= 500; ix += 20)
+        {
+             KeStallExecutionProcessor(20);
+             PortSC.AsULONG = READ_REGISTER_ULONG(PortStatusReg);
+
+             DPRINT("EHCI_RH_PortResetComplete: Reset port - %x\n", Port);
+
+             if (PortSC.PortReset == 0)
+                 break;
+        }
+    }
+    while (PortSC.PortReset == 1 && (PortSC.AsULONG != -1));
```

Comment:
`(PortSC.AsULONG != -1)` - this one looks a bit suspicious.

Can ULONG be less than zero? (u is for unsigned)